src/playbin.js

```diff
@@ -0,0 +1,157 @@
+/* exported Playbin */
+
+const {Gdk, GLib, GObject, Gtk, Gst} = imports.gi;
+
+var Playbin = GObject.registerClass({
+    CssName: 'playbin',
+    Properties: {
+        uri: GObject.ParamSpec.string('uri', 'URI', '',
+            GObject.ParamFlags.READWRITE,
+            ''
+        ),
+    },
+    Signals: {
+        clicked: {},
+        start: {},
+        done: {},
+    },
+}, class Playbin extends Gtk.EventBox {
+    _init(props) {
+        super._init(props);
+
+        this._style = this.get_style_context();
+        this._css_provider = new Gtk.CssProvider();
+        this._style.add_provider(
+            this._css_provider,
+            Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION
+        );
+
+        this.connect('button-release-event', () => {
+            this.emit('clicked');
+            return Gdk.EVENT_PROPAGATE;
+        });
+    }
+
+    _ensurePlaybin () {
+        if (this._playbin)
+            return;
+
+        Gst.init_check(null);
+
+        this._playbin = Gst.parse_launch('playbin3 name=playbin video-sink=gtksink');
+        this._playbin.videoFilter = Gst.parse_bin_from_description(
+            'videocrop name=videocrop ! alpha method=green',
+            true
+        );
+        this._videocrop = this._playbin.videoFilter.get_by_name('videocrop');
+
+        this._video_widget = this._playbin.videoSink.widget;
+        this._video_widget.expand = true;
+        this._video_widget.noShowAll = true;
+        this._video_widget.ignoreAlpha = false;
+        this._video_widget.hide();
+        this.add(this._video_widget);
+
+        this._video_widget.connect('size-allocate', this._updateCropArea.bind(this));
+
+        this._playbin.get_bus().add_watch(0, this._bus_watch.bind(this));
+    }
+
+    _updateCropArea () {
+        const width = this._video_widget.get_allocated_width();
+        const height = this._video_widget.get_allocated_height();
+
+        if (width < 2 || height < 2 || !this._video_width || !this._video_height)
+            return;
+
+        const crop_x = (this._video_width - width) / 2;
+        const crop_y = (this._video_height - height) / 2;
```

Comment:
Is it worth it here to check whether the crop_x/crop_y values are already the same set on the videocrop and bail our early, or does videocrop already optimize that?

Reply:
It already optimizes that.